torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1438,6 +1438,116 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   return work;
 }
 
+template <typename Fn, typename PreProcess, typename PostProcess>
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::oneAndAll(
+    std::vector<at::Tensor>& tensor,
+    std::vector<std::vector<at::Tensor>>& tensor_list,
```

Comment:
Do we really need a new method like this that has a lot of similarities to `collective`? What if we just make `collective` take a `vector<vector<Tensor>>` instead of just `vector<Tensor>`? Then modify current code that passes in a `vector<Tensor>` to pass in a `vector<vector<Tensor>>` of size 1?

Reply:
Yeah I tried to consolidate all into `collective` before introducing this oneAndAll, failed mainly because for `gather` and `scatter`, the inputs and outputs are different (they are swapped), so need someway to distinguish it (i.e. we only can record input stream on one rank for `scatter`, not all ranks as its only available on rootRank, same for `gather`), so it was pretty hard to generalize. Let me try again and see if this could work

Reply:
I actually found a way to work around this issue, by not introducing `oneAndAll`, and still reuse `collective` to do the job, the trick is to ignore the inputs/outputs specified in `fn` and use function capture to deal with `tensor` and `tensor_list` directly. Updated the PRs to reflect the changes. Let me know if this works :)